### PR TITLE
Package Release History draft

### LINKED ISSUE #1587: Implement a Package Release History Page
# Task

Add a page that lists all the released versions of a package.

Screenshot from 2023-10-06 10-42-54

Here is [a Figma document with the design for the versions page of the package](https://www.figma.com/file/J0w4b9exaHhVpvabKoGqTn/Design-Github-Issues?type=design&node-id=34%3A203&mode=design&t=AsJdiTX8o4t3ALZ7-1).

How to solve this issue:

1. Create a new template `package_versions.eml` that renders the page in the linked Figma design. Take inspiration on how to iterate over lists from other templates. Be careful about indentation, the templating engine used in this project is sensitive to it.
2. Add a rule to the `dune` file at `src/ocamlorg_frontend/dune` to process the `.package_versions.eml` template into an OCaml module `package_versions.ml`.
3. Add a function for the new template to `ocamlorg_frontend.ml`.
4. Add a handler `package_versions` to `handler.ml` that gets the relevant data (package and versions) and renders the new template.
5. Add a route for the new package versions page to `src/global/url.ml` and register the handler from the previous step it in the router at `router.ml`.
6. In the package area, on the package overview page at https://ocaml.org/p/irmin/latest (template `package_overview.eml`), a link needs to be added to go to the new package versions page. Hint: The little icon is `Icons.tag`.
7. 
Screenshot 2023-10-06 at 10-40-55 irmin 3 8 0 (latest) · OCaml Package

### PR DESCRIPTION
**Pull request by kate-bagenzo**
Resolves #1587.

Here's where I am with #1587. A couple of notes/obvious issues, besides the page missing styles:

- A URL is generated for every version of a package (instead of the URL being at `dune/versions` it is at `dune/3.11/versions` and `dune/2.92/versions` and so on). Changing `url.ml` and 'router.ml' was tricky - I was mostly copying other entries to get something workable - and it doesn't seem to accept simply taking the version out.
- No number of package versions displayed
- A documentation link is displayed regardless of whether or not documentation exists
- Release date is for the first version only

### CODE REVIEW

**Comment by sabine** [comment]
Hey @kate-bagenzo, that's a great start, here's some pointers:

1. I think it makes sense to either (a) give the new page a canonical URL of `/p/PACKAGE/latest/versions`, or (b) change the route to be `/p/PACKAGE/versions`.
2. there needs to be a link from `package_overview.eml` going to the new page.

Apart from that, this is nice and the challenge that's left is to create HTML and CSS styles.

File: src/ocamlorg_frontend/css/styles.css [hunk 1]
```diff
@@ -96,42 +96,47 @@ body {
 }
 
 .code-preview {
-  font-family: Roboto Mono,ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,Liberation Mono,Courier New,monospace;
```
Comment (sabine) [inline src/ocamlorg_frontend/css/styles.css]: would prefer unrelated formatting changes to appear in a different pull request in favor of keeping the diff small on this one

File: src/ocamlorg_frontend/css/styles.css [hunk 2]
```diff
@@ -218,40 +223,75 @@ summary::marker {
...
+.tr-selected:first-of-type td .rounded-full {
+  background-color: #F36528;
+  height: 0.75rem;
+  width: 0.75rem;
+  margin: 0;
+}
+
+.tr-selected:last-of-type td div div {
+  display: none;
+}
```
Comment (sabine) [inline src/ocamlorg_frontend/css/styles.css]: this should probably not be here as you can make an if then else in the templates when rendering tailwind styles

File: src/ocamlorg_frontend/pages/package_versions.eml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            <tr class="py-4">
+                <th class="text-left font-normal border-b-2 border-gray-200 py-4 pr-12 pl-3"><%s! Icons.bars_3 "h-5 w-5"; %></th>
+                <th class="text-left font-normal border-b-2 border-gray-200 pr-12">Version</th>
+                <th class="text-left font-normal border-b-2 border-gray-200 pr-12">Release Date</th>
+                <th class="text-left font-normal border-b-2 border-gray-200 pr-12 pl-5">Links</th>
+            </tr>
+        </thead>
+        <tbody>
+          <% package.versions |> List.iter (fun (version : string) -> %>
+          <tr class="pt-5 tr-selected">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/package_versions.eml]: Never do padding on surrounding elements when you can do the padding on the `<a>` tag.

The click area of a link should always be as big as the visuals of the link.

When there's white space around links, it's always worth to consider if that can be subsumed into the link's padding so that the click area is bigger - assume that some people are very clumsy and have a hard time clicking/tapping on things or they just have big fingers.

In this particular case, the padding should be on the link.

**Review by sabine — COMMENTED** [review]
Great, this is coming together! :+1: Here's a quick review, I didn't check all the details on the Tailwind styles, but will next time I review this.

Here are some tips to how to get closer to the design:

- You can enable "dev mode" in Figma to more easily read the CSS values from the designs, and to better see the spacings between different elements:
Screenshot from 2023-10-24 15-37-43


Here's some more feedback:

1. In the Figma design, the size of the circle back button on the page is sized slightly different:
image
You can see that the design says that the size is `2.8125rem`. The closest Tailwind value to this is `11` instead of `12`, which suggests that `h-11 w-11` should be used instead of `h-12 w-12`. (For Tailwind values see https://tailwindcss.com/docs/customizing-spacing#default-spacing-scale.) Please also double check the size of the arrow.

1. It would be great to list the number of published versions in brackets. That is, instead of "All versions" have "Versions (20)" if 20 is the total number of published versions.
image

1. check font weight and sizes. See https://tailwindcss.com/docs/font-weight for available font weights in Tailwind. For example, here `font-bold` is used when the design selects `font-medium`:
image

File: src/ocamlorg_frontend/css/styles.css (hunk 1 above)
Comment (kate-bagenzo) [reply src/ocamlorg_frontend/css/styles.css]: Oops! Code formatter was enabled by accident - ignore.

File: src/ocamlorg_frontend/pages/package_versions.eml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+              </div>
+            </td>
+            <td class="text-left bg-primary-200">
+              <a href="/p/<%s package.name %>/<%s version %>" class="py-5 mr-10 hover:no-underline hover:text-primary-600 focus:text-primary-600 focus:no-underline font-bold focus:font-bold text-base"><%s version %></a>
+            </td>
+            <td class="pr-10 text-left text-sm bg-primary-200">
+              <%s Utils.human_date_of_timestamp package.publication %>
+            </td>
+            <td class="text-left text-sm bg-primary-200 rounded-r">
+              <a href="/p/<%s package.name %>/<%s version %>/doc/index.html" class="text-primary-600 flex items-center mx-4 font-medium"><%s! Icons.document "h-5 w-5"; %> Documentation</a>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/package_versions.eml]:
```suggestion
              <a href="/p/<%s package.name %>/<%s version %>/doc/index.html" class="text-primary-600 flex items-center p-4 font-medium"><%s! Icons.document "h-5 w-5"; %> Documentation</a>
```

File: src/ocamlorg_frontend/pages/package_versions.eml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  <div class="container-fluid flex items-center gap-2">
+    <a href="/p/<%s package.name %>" class="h-11 w-11 m-5 border-primary-600 border-2 rounded-full flex items-center justify-center"><%s! Icons.arrow_left "h-6 w-6 text-primary-600"; %></a>
+    <h1 class="text-2xl font-medium"><%s package.name %> Versions </h1><div class="text-2xl">(<%i List.length package.versions %>)</div>
+  </div>
+  <div class="container-fluid flex justify-center">
+    <table class="table-auto border-separate border-spacing-y-6">
+        <thead class="container-fluid text-md text-lighter text-left">
+            <tr class="py-4">
+                <th class="text-left font-normal border-b-2 border-gray-200 py-4 pr-14 pl-3"><%s! Icons.bars_3 "h-5 w-5"; %></th>
+                <th class="text-left font-normal border-b-2 border-gray-200 pr-12">Version</th>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/package_versions.eml]:
```suggestion
                <th class="text-left font-normal border-b-2 border-gray-200 pl-5 pr-12">Version</th>
```

File: src/ocamlorg_frontend/pages/package_versions.eml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+          <tr class="tr-selected">
+            <% if List.nth package.versions 0 = version then ( %>
+            <td class="text-left font-normal py-5 px-4 bg-primary-200 rounded-l">
+              <div class="bg-primary-600 h-3 w-3 rounded-full"></div>
+              <div class="absolute h-16 pl-1">
+                <div class="-ml-px w-1 border-solid border-r-2 border-gray-200 h-full"></div>
+              </div>
+            </td>
+            <td class="text-left bg-primary-200">
+              <a href="/p/<%s package.name %>/<%s version %>" class="py-5 mr-10 hover:no-underline hover:text-primary-600 focus:text-primary-600 focus:no-underline font-bold focus:font-bold text-base"><%s version %></a>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/package_versions.eml]:
```suggestion
              <a href="/p/<%s package.name %>/<%s version %>" class="p-5 mr-10 hover:no-underline hover:text-primary-600 focus:text-primary-600 focus:no-underline font-bold focus:font-bold text-base"><%s version %></a>
```

**Review by sabine — COMMENTED** [review]
Looks really great overall now, the only thing I can find after these minor suggestions are applied is the collapsing on smallest screen sizes:

Screenshot 2023-10-26 at 10-43-36 irmin Versions

It seems that the paddings/margins that work well at higher screen sizes do not work well here. It may make sense to use `sm:` or `md:` prefixes to assign different padding/margin based on screen width.

**Review by sabine — COMMENTED** [review]
I have some changes here that improve the collapsing of the package versions table on small screens:

https://github.com/kate-bagenzo/ocaml.org/compare/main...sabine:ocaml.org:package_versions?expand=1

Would love for them to be included in the PR. (I can't push to your branch because it's your protected main branch.)

File: src/ocamlorg_frontend/css/styles.css (hunk 1 above)
Comment (sabine) [reply src/ocamlorg_frontend/css/styles.css]: same problem again, code formatter changed the formatting of styles.css

File: src/ocamlorg_frontend/pages/package_versions.eml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+              <div class="bg-primary-600 h-3 w-3 rounded-full"></div>
+              <div class="absolute h-[4.5rem] pl-1">
+                <div class="-ml-px w-1 border-solid border-r-2 border-gray-200 h-full"></div>
+              </div>
+            </td>
+            <td class="text-left bg-primary-200">
+              <a href="/p/<%s package.name %>/<%s version %>" class="p-5 hover:no-underline hover:text-primary-600 focus:text-primary-600 focus:no-underline font-bold focus:font-bold text-base md:mr-10"><%s version %></a>
+            </td>
+            <td class="pr-4 text-left text-sm bg-primary-200">
+              <%s Utils.human_date_of_timestamp package.publication %>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/package_versions.eml:35]: Sorry, just noticed: we need the publication date of that particular version of the package here. This is rendering the publication date of the latest version of the package, for every version.

Comment (kate-bagenzo) [reply src/ocamlorg_frontend/pages/package_versions.eml:35]: as first mentioned on this PR, i'm completely stuck on this issue - any advice?

Comment (sabine) [reply src/ocamlorg_frontend/pages/package_versions.eml:35]: Ah sorry, I overlooked that. To render the publication dates of the other versions, we need to look them up in `handler.ml` and pass them to the template as an argument.

This looks like it will require some changes in `ocamlorg_package.ml` and `ocamlorg_package.mli`. For example, we could add a function `get_versions_with_publication_date : state -> string -> (Version.t, float) list`, which, given the package state and the name of a package returns a list of all the versions and respective publication dates of the package.

Comment (kate-bagenzo) [reply src/ocamlorg_frontend/pages/package_versions.eml:35]: ``get_versions`` in ``ocamlorg_package.ml`` seems pretty close to what i need but i'm not sure how to get the publication dates

Comment (sabine) [reply src/ocamlorg_frontend/pages/package_versions.eml:35]: yes, for the new `get_versions_with_publication_date` function, I would start with what `get_versions` does and then try to change that to also give the publication date `float` value.

Comment (kate-bagenzo) [reply src/ocamlorg_frontend/pages/package_versions.eml:35]: Right, that's what my question was. Do you have any more specific guidance?

Comment (sabine) [reply src/ocamlorg_frontend/pages/package_versions.eml:35]: You should also look at the function above `get_versions`. It is called `get_by_name` and retrieves the name, version, and package info for a package. The package info contains the publication date.

As a hint: the function `Version.Map.bindings` returns a list of key-value pairs `(version, info)`.

**Comment by sabine** [comment]
Hey @kate-bagenzo, no reason to close this. I'll finish the remaining parts.
